neutone_sdk/conv.py

```diff
@@ -0,0 +1,365 @@
+import logging
+import os
+from typing import Optional, Tuple, Union
+
+import torch as tr
+import torch.nn.functional as F
+from torch import Tensor
+from torch import nn
+
+logging.basicConfig()
+log = logging.getLogger(__name__)
+log.setLevel(level=os.environ.get('LOGLEVEL', 'INFO'))
+
+
+class PaddingCached(nn.Module):
+    def __init__(self,
+                 n_ch: int,
+                 padding_l: int,
+                 padding_r: int = 0,
+                 use_dynamic_bs: bool = True,
+                 batch_size: int = 1,
+                 debug_mode: bool = True) -> None:
+        """
+        Cached padding for cached convolutions. Handles dynamic batch sizes by default
+        at the expense of dynamic memory allocations.
+
+        Args:
+            n_ch: Number of channels.
+            padding_l: Number of padding samples on the left side.
+            padding_r: Number of padding samples on the right side.
+            use_dynamic_bs: If True, the padding will dynamically change batch size to
+                            match the input.
+            batch_size: If known, the initial batch size can be specified here to avoid
+                        dynamic changes.
+            debug_mode: If True, assert statements are enabled.
+        """
+        super().__init__()
+        if debug_mode:
+            assert n_ch > 0
+            assert batch_size > 0
+        self.n_ch = n_ch
+        self.padding_l = padding_l
+        self.padding_r = padding_r
+        self.use_dynamic_bs = use_dynamic_bs
+        self.debug_mode = debug_mode
+        self.register_buffer("pad_l_buf", tr.zeros((batch_size, n_ch, padding_l)))
+        self.register_buffer("pad_r_buf", tr.zeros((batch_size, n_ch, padding_r)))
+
+    def reset(self, batch_size: Optional[int] = None) -> None:
+        if batch_size is not None:
+            self.pad_l_buf = self.pad_l_buf.new_zeros(
+                (batch_size, self.n_ch, self.padding_l))
+            self.pad_r_buf = self.pad_r_buf.new_zeros(
+                (batch_size, self.n_ch, self.padding_r))
+        else:
+            self.pad_l_buf.zero_()
+            self.pad_r_buf.zero_()
+
+    def prepare_for_inference(self) -> None:
+        self.debug_mode = False
+        self.reset()
```

Comment:
Below you have a comment saying torchscript needs to pass in the batch size here, but this one is fine?

Reply:
Yeah `prepare_for_inference` isn't converted to torchscript so it's fine to omit the optional argument. I'll add a comment to clarify this.